heron/uploaders/src/java/com/twitter/heron/uploader/gcs/GcsUploader.java

```diff
@@ -0,0 +1,225 @@
+
+  private GcsController gcsController;
+  private String bucket;
+  private File topologyPackageFile;
+  private String topologyObjectName;
+
+  // Stores the path to the backup version of the topology in case the deploy fails and
+  // it needs to be undone. This is the same as the existing path but prepended with `previous-`.
+  // This serves as a simple backup incase we need to revert.
+  private String previousTopologyObjectName;
+
+  @Override
+  public void initialize(Config config) {
+    bucket = GcsContext.getBucket(config);
+
+    if (Strings.isNullOrEmpty(bucket)) {
+      throw new RuntimeException("Missing heron.uploader.gcs.bucket config value");
+    }
+
+    // get the topology package location
+    final String topologyPackageLocation = Context.topologyPackageFile(config);
+    final String topologyName = Context.topologyName(config);
+
+    topologyPackageFile = new File(topologyPackageLocation);
+    final String topologyFilename = topologyPackageFile.getName();
+    topologyObjectName = generateStorageObjectName(topologyName, topologyFilename);
+    previousTopologyObjectName =
+        generateStorageObjectName(topologyName, BACKUP_PREFIX + topologyFilename);
+
+    gcsController = createGcsController(config, bucket);
+  }
+
+  @Override
+  public URI uploadPackage() throws UploaderException {
+    // Backup any existing files incase we need to undo this action
+    final StorageObject previousStorageObject = gcsController.getObject(topologyObjectName);
+    if (previousStorageObject != null) {
+      try {
+        gcsController.copyObject(topologyObjectName, previousTopologyObjectName,
+            previousStorageObject);
+      } catch (IOException ioe) {
+        throw new UploaderException("Failed to back up previous topology version", ioe);
+      }
+    }
+
+    final StorageObject storageObject;
+    try {
+      storageObject = gcsController.createObject(topologyObjectName, topologyPackageFile);
+    } catch (IOException ioe) {
+      throw new UploaderException(
+          String.format("Error writing topology package to %s %s",
+              bucket, topologyObjectName), ioe);
+    }
+
+    final String downloadUrl = getDownloadUrl(bucket, storageObject.getName());
+    LOG.info("Package URL: " + downloadUrl);
+    try {
+      return new URI(downloadUrl);
+    } catch (URISyntaxException e) {
+      throw new UploaderException(
+          String.format("Could not convert URL %s to URI", downloadUrl), e);
+    }
+  }
+
+  @Override
+  public boolean undo() {
+    // Try to get the previous version. This will be null on the first deploy.
+    final StorageObject previousObject = gcsController.getObject(previousTopologyObjectName);
+    if (previousObject != null) {
+      try {
+        gcsController.copyObject(previousTopologyObjectName, topologyObjectName, previousObject);
+      } catch (IOException ioe) {
+        LOG.log(Level.SEVERE, "Reverting to previous topology version failed", ioe);
```

Comment:
Can you fetch the version from the StorageObject? Are there even versions that would be useful for a user to see?